### ISSUE TITLE
Incorrect Decoder Compilation for DEIMv2 on neuronx

### ISSUE THREAD

**Issue by cricket1** [aws-neuron/aws-neuron-samples#116]
## Summary

The AWS Neuron compiler (`torch-neuronx`) produces incorrect outputs when compiling a transformer-based object detection decoder with discrete attention mechanism. The compiled model produces vastly different logits and bounding boxes compared to PyTorch, resulting in zero detections on real images.

## Environment

```bash
# System info
OS: Ubuntu (Linux 6.8.0-1040-aws)
Instance: inf2.xlarge (tested on CPU instance for compilation)

# Python packages (confirmed versions)
Python: 3.10.12
torch: 2.8.0
torch-neuronx: 2.8.0.2.10.16998+e9bf8a50
torch-xla: 2.8.1
neuronx-cc: 2.21.33363.0+82129205
torchvision: 0.23.0
```

### Expected Behavior
When compiling a transformer decoder with discrete attention to Neuron format, the compiled model should produce outputs numerically close to PyTorch (within typical BF16 precision tolerance of ~1e-2).

### Actual Behavior
The Neuron-compiled decoder produces completely incorrect outputs:
- **Logits max difference**: 7.84 (compared to ~1e-2 expected)
- **Logits mean difference**: 1.25 (compared to ~1e-4 expected)
- **Relative difference**: 160,087% (!)
- **Detection count**: 0 detections (PyTorch: 60 detections)
- **Top-5 confidence scores**: [0.25, 0.24, 0.23, 0.21, 0.21] (PyTorch: [0.93, 0.90, 0.89, 0.84, 0.84])

### Components Affected
- **Backbone (HGNetv2)**: ✓ Works correctly (max diff ~0.7)
- **Encoder (HybridEncoder)**: ✓ Works correctly (max diff ~0.18)
- **Decoder (DEIMTransformer with discrete attention)**: ✗ Completely broken

### 1. Model Architecture

The decoder is a DETR-style transformer decoder with:
- Multi-scale deformable attention using **discrete indexing** (not F.grid_sample)
- 3 decoder layers
- 300 object queries
- 80 classes (COCO)
- 2 feature levels

**Key detail**: The model uses `cross_attn_method: 'discrete'` which implements deformable attention via discrete tensor indexing instead of `F.grid_sample()` (which is not supported by Neuron).

### 3. Full Reproduction

Complete reproduction code available at: `/home/ubuntu/sh_deimv2/`

Key files:
- `convert_components_neuronx.py`: Conversion script ([View on GitHub Gist](https://gist.github.com/cricket1/65c14f61644d312532172f1e9939b252))
- `verify_neuronx_on_image.py`: Verification on real image ([View on GitHub Gist](https://gist.github.com/cricket1/5aa2c3fb2611e2cde983262ec8f827d1))
- `compare_all_methods.py`: Comprehensive comparison ([View on GitHub Gist](https://gist.github.com/cricket1/e03f92661eb1b2773a954ade95046b70))

To reproduce:
```bash
# Convert decoder to Neuron
python convert_components_neuronx.py \\
    --checkpoint models/deimv2_hgnetv2_n_coco.pth \\
    --config configs/deimv2/deimv2_hgnetv2_n_coco.yml \\
    --component all

# Verify (shows the bug)
python verify_neuronx_on_image.py --image example.jpg
```

### What We've Tried

1. ✗ **Different compiler optimizations**: Tested `-O0`, `-O1`, `-O2`
2. ✗ **FP32 precision**: `--fp32-cast=all`, `--fp32-cast=matmult`
3. ✗ **Disabled auto-casting**: `--auto-cast=none`
4. ✗ **Model type hint**: `--model-type=transformer`
5. ✗ **Various combinations**: See `try_neuronx_compiler_settings.py`

None of the above produced correct outputs.

### Key Observations

1. **Backbone and encoder compile correctly** - The issue is isolated to the decoder
2. **PyTorch discrete attention works perfectly** - The implementation is correct in PyTorch
3. **Error is not precision-related** - Differences are 1000x larger than BF16 tolerance
4. **Consistent failure** - Fails on both random inputs and real images

**Comment by 0x6b64** [comment]
Hi @cricket1 , I used the latest release container public.ecr.aws/neuron/pytorch-inference-neuronx:2.8.0-neuronx-py311-sdk2.26.1-ubuntu22.04 with python3.11 && the same software versions mentioned above. 

One change I had to make was in deimv2_hgnetv2_n_coco.yml to specify `cross_attn_method: 'discrete'`.

I'll attach the full compilation report shortly but here is the summary:

```
======================================================================
Verifying Decoder
======================================================================
Running PyTorch version...
Running Neuron version...

Output[0] comparison:
  Shape: torch.Size([1, 300, 80])
  Max diff: 5.92e+00
  Mean diff: 1.10e+00
  Tolerance: 1.00e-02
  ⚠ Exceeds tolerance (this is often acceptable)
  PyTorch sample: [-3.0395820140838623, -5.233051776885986, -4.355584144592285, -8.036619186401367, -3.838811159133911]
  Neuron sample: [-3.6187691688537598, -4.961472988128662, -6.1292243003845215, -5.711883068084717, -4.57215690612793]

Output[1] comparison:
  Shape: torch.Size([1, 300, 4])
  Max diff: 9.48e-01
  Mean diff: 1.65e-01
  Tolerance: 1.00e-02
  ⚠ Exceeds tolerance (this is often acceptable)
  PyTorch sample: [0.47292083501815796, 0.4709470868110657, 1.0251448154449463, 0.9928902387619019, 0.47625264525413513]
  Neuron sample: [0.5242082476615906, 0.15176597237586975, 1.156178593635559, 0.8296037316322327, 0.6077398061752319]

✓ Decoder verification PASSED

======================================================================
CONVERSION SUMMARY
======================================================================
Backbone       : ✓ PASSED
Encoder        : ✓ PASSED
Decoder        : ✓ PASSED

Output files:
  decoder_neuronx.pt               4.89 MB
  encoder_neuronx.pt               1.82 MB
  backbone_neuronx.pt             14.85 MB
```

This is the verification script output:

```
======================================================================
NeuronX Component Verification on Real Image
======================================================================

Loading PyTorch model from: deimv2_hgnetv2_n_coco.pth
        ### Backbone.act: relu ###     
        ### Backbone.act: relu ###     
     --- Use Gateway@True ---
     --- Use Share Bbox Head@False ---
     --- Use Share Score Head@False ---
     --- Wide Layer@1 ---
✓ PyTorch model loaded
Input size: (640, 640)

Loading image: example.jpg
Original size: 500x375
Input tensor shape: torch.Size([1, 3, 640, 640])

######################################################################
# LOADING NEURONX COMPONENTS
######################################################################
Loading NeuronX components from: neuronx_components
✓ All NeuronX components loaded

######################################################################
# RUNNING INFERENCE
######################################################################

PyTorch inference...
NeuronX inference...
✓ Inference complete

######################################################################
# RAW OUTPUT COMPARISON
######################################################################

  Logits:
    Shape:        (1, 300, 80)
    Max Diff:     7.836291e+00
    Mean Diff:    1.253349e+00
    Max Rel Diff: 160087.07%

  Boxes:
    Shape:        (1, 300, 4)
    Max Diff:     8.710204e-01
    Mean Diff:    1.343510e-01
    Max Rel Diff: 1784.42%

######################################################################
# POST-PROCESSING AND DETECTION COMPARISON
######################################################################

======================================================================
DETECTION COMPARISON
======================================================================

Number of detections:
  PyTorch: 60
  Neuron:  0

Top 5 Detections:
  #    Source     Class                Score      Box (x1,y1,x2,y2)             
  ---- ---------- -------------------- ---------- ------------------------------
  1    PyTorch    person               0.934      [119.3,42.6,165.2,133.0]
  2    PyTorch    person               0.901      [370.0,67.2,432.3,118.1]
  3    PyTorch    donut                0.890      [77.4,200.9,118.3,228.3]
  4    PyTorch    donut                0.844      [195.0,181.2,228.6,198.4]
  5    PyTorch    person               0.839      [334.2,46.9,370.8,118.5]


######################################################################
# CREATING VISUALIZATION
######################################################################

✓ Comparison visualization saved to: verification_comparison.jpg

======================================================================
VERIFICATION SUMMARY
======================================================================
  Image: example.jpg (500x375)
  PyTorch detections: 60
  NeuronX detections: 0
  Confidence threshold: 0.45

  ✓ NeuronX components are working correctly!
  ✓ Small differences in scores/boxes are expected due to compilation optimizations
  ✓ Visual comparison saved to verification_comparison.jpg
nrtucode: internal error: 54 object(s) leaked, improper teardown
```

@cricket1 Do you have a hard dependency on python3.10? I'll rerun testing with 3.10, but from the scripts you provided, it seems like we are passing the verification checks (using python3.11).

**Comment by cricket1** [comment]
@0x6b64 could you make any progress

**Comment by parmoham-aws** [comment]
We're still debugging the issue and will provide updates as we make progress. Thanks for your patience.

**Comment by ivashkst** [comment]
Thank you for your patience. We are still looking into this issue.

**Comment by tapansahni7** [comment]
Hi, thank for your patience. We have fixed this issue in the latest Neuron SDK release 2.32.0. Please let me us know if there are any questions or issues.
